### ISSUE TITLE
Implement basic accessibility service

### ISSUE THREAD

**Issue by jdm** [servo/servo#4344]
We should build the infrastructure necessary to work with native accessibility services (eg. screenreaders). I imagine this being set up like the devtools; the code for communicating with the external service can exist in its own crate, with a channel per global to obtain information about the DOM and ARIA-related features. The sooner we do this basic plumbing, the harder it will be to screw it up down the road as Servo gets larger.

We can use Gecko's code as a reference here:
http://mxr.mozilla.org/mozilla-central/source/accessible/atk/Platform.cpp#124 (linux)
http://mxr.mozilla.org/mozilla-central/source/accessible/mac/mozAccessibleProtocol.h?force=1 (mac)
http://mxr.mozilla.org/mozilla-central/source/widget/cocoa/nsChildView.mm#6617 (mac)

**Comment by jdm** [comment]
Previous investigative work I did:
https://github.com/jdm/glutin/tree/a11y - branch of glutin that adds the necessary platform integration with OS X's accessibility services
https://github.com/jdm/accessible - crate designed to provide platform-independent accessible representations for Rust programs; OS X support present, very preliminary Linux work

I hear that Windows is even more complicated.

**Comment by karan1276** [comment]
This project looks interesting. I am looking for a slightly biggish project, but right now i only know the basics. Would it be a good idea for me to try this?

**Comment by jdm** [comment]
No. There's no information available on how to do this correctly.

**Comment by karan1276** [comment]
okay

**Comment by karan1276** [comment]
I did some research, gecko [implements this](https://developer.mozilla.org/en-US/docs/Mozilla/Accessibility/Accessibility_architecture) in a good way. Webkit is also doing something very similar.

There are 2 layers:
- Platform dependent - since all platforms have their own API's we write this layer so  we get a consistent interface on all platform.
- Platform Independent layer - This layer maintains it's own accessibility tree and provides methods to access it.


Accessibility is something that doesn't matter to most people, but to those it does matter, it matters a lot!. It would be nice to have servo support accessibility. 

I think this project is really interesting and kind of hard (lots of research needed) and makes a perfect candidate for GSoC ? It may be too huge to do all this in one go. So i would like suggest something.

I would like to implement the platform dependent layer complete, so we can lay a good foundation for future work. That way implementing this will become easier. In fact i would suggest we separate that to a different crate so any application needing a cross-platform accessibility API will be able to use it.

@jdm i would love to know what you think about it? maybe you have another plan for this, which is completely fine. Then we can just drop this idea.

**Comment by jdm** [comment]
The separation between platform-independent and platform-dependent layers is something that I started creating in https://github.com/jdm/accessible. There are many aspects of this that suggest it would be a good GSoC project, unfortunately we don't have anybody with experience implementing accessibility integration who could act as a mentor. This means finding someone willing to take the time to be a mentor and learn all about it, and that requires more time than I can commit to.

**Comment by karan1276** [comment]
Okay, ill drop this for now. Maybe later when someone has the leisure of time.

**Comment by pshaughn** [comment]
WPT expects ARIA to at least exist on the DOM level even if it the test harness has no way of looking at the UI-facing aspects; dom/nodes/aria-attribute-reflection.tentative.html and dom/nodes/aria-element-reflection.tentative.html each test a long list of ARIA semantic features.

**Comment by Keithcat1** [comment]
I built Servo about a year and a half ago and because I'm a screen reader user, it turned out that the only accessible and usable part of the program was the address dialog, so I wasn't able to use it. Thanks for the project though, it sounds exciting!

**Comment by jdm** [comment]
The right path forward is probably to learn about the https://crates.io/crates/accesskit and how to integrate it with servo, since it would handle the platform-specific integration.

**Comment by Keithcat1** [comment]
Yes, Egui already uses that and it works beautifully.

**Comment by jdm** [comment]
I think there are a few big parts to this work:
* supporting screenreaders for servoshell
* exposing an embedding API to obtain an accessibility tree for the contents of a servo webview
* supporting subtrees in accesskit (https://github.com/AccessKit/accesskit/issues/531#issuecomment-2841694160)

egui already supports accesskit integration for built-in components, and I tried to enable it in [this branch](https://github.com/servo/servo/compare/main...jdm:servo:a11y?expand=1) based on the way accesskit is [used by eframe](https://github.com/search?q=repo%3Aemilk%2Fegui+path%3A%2F%5Ecrates%5C%2Feframe%5C%2F%2F+access&type=code). These changes were not enough to make the macOS Accessibility Inspector recognize the Servo app window, unfortunately, and I ran out of ideas.

**Comment by simonwuelker** [comment]
> I think there are a few big parts to this work:
> 
>     * supporting screenreaders for servoshell
> 
>     * exposing an embedding API to obtain an accessibility tree for the contents of a servo webview
> 
>     * supporting subtrees in accesskit ([Wrap accessibility tree in Window node by default AccessKit/accesskit#531 (comment)](https://github.com/AccessKit/accesskit/issues/531#issuecomment-2841694160))
> 
> 
> egui already supports accesskit integration for built-in components, and I tried to enable it in [this branch](https://github.com/servo/servo/compare/main...jdm:servo:a11y?expand=1) based on the way accesskit is [used by eframe](https://github.com/search?q=repo%3Aemilk%2Fegui+path%3A%2F%5Ecrates%5C%2Feframe%5C%2F%2F+access&type=code). These changes were not enough to make the macOS Accessibility Inspector recognize the Servo app window, unfortunately, and I ran out of ideas.

That's odd, because your branch works on fedora (the screenreader activates when hovering the servoshell window and says "frame").

**Comment by simonwuelker** [comment]
https://github.com/AccessKit/accesskit/issues/465#issuecomment-2408973079:

> We discovered that for an application to show up in the inspector, it needs to be packaged and run as a macOS app bundle, not simply run from the terminal.

That's probably the issue.

**Comment by jdm** [comment]
> [AccessKit/accesskit#465 (comment)](https://github.com/AccessKit/accesskit/issues/465#issuecomment-2408973079):
> 
> > We discovered that for an application to show up in the inspector, it needs to be packaged and run as a macOS app bundle, not simply run from the terminal.
> 
> That's probably the issue you're running into.

Oh, thanks for finding that. That's... fun.

**Comment by simonwuelker** [comment]
Do you want to open a PR for your branch?

**Comment by jdm** [comment]
Yeah, I'll do that tonight.

**Comment by simonwuelker** [comment]
Chromium has very useful documentation about how their accessiblity service works, maybe we can make use of some of that:

* https://chromium.googlesource.com/chromium/src/+/HEAD/docs/accessibility/overview.md
* https://chromium.googlesource.com/chromium/src/+/HEAD/docs/accessibility/browser/how_a11y_works.md
* https://chromium.googlesource.com/chromium/src/+/HEAD/docs/accessibility/browser/how_a11y_works_2.md
* https://chromium.googlesource.com/chromium/src/+/HEAD/docs/accessibility/browser/how_a11y_works_3.md

I believe everything that happens in chromiums browser process would already be handled by accesskit and the platform-specific adapters.

**Comment by tapper82** [comment]
Hi I am looking forward to being able to test this out on windows when it is ready. I use NVDA. When anything is ready for running on windows pleas ping me. I cant code, but I am a power user of screen readers.

**Comment by jdm** [comment]
@tapper82 That's great to hear! We'll be sure to tag you when we have something that can be tested.

**Comment by nebeker** [comment]
@jdm  gotcha, checking that issue now. I'm surprised it didn't pop up under the accessibility label. Haha.

Is there any high level info on the roadmap I can check out?

**Comment by lukewarlow** [comment]
There's no explicit roadmap at the minute that I'm aware of, but we're actively working on this issue. So far we've been evaluating how best to integrate with the platform accessibility APIs in a way that works best for Servo's design. AccessKit with some modifications is what we've started integrating.
 https://www.igalia.com/2025/10/09/Igalia,-Servo,-and-the-Sovereign-Tech-Fund.html

**Comment by jdm** [comment]
Relevant explorations so far:
* https://github.com/servo/servo/pull/41173
* https://github.com/AccessKit/accesskit/pull/655
